File: frontend/src/coverage.ts
// Pure logic for the coverage heatmap (M2-07c): the geotile arithmetic that
// turns a cell key into a box, the log-scale color ramp, and the two rules
// that decide what the legend says. No map/DOM access here — `mapLayers.ts`
// and the components call these and do the drawing.

import type { CoverageCell, CoverageCompleteness, CoverageResponse } from './api';
import type { Bbox } from './types';

export class InvalidCellKey extends Error {}

// The same bound as `catalog.coverage.MAX_GEOTILE_LEVEL`: geotile runs out
// here (below a metre a cell is past any imagery), and Earth Search happens
// to draw the same line (a precision outside 0..29 comes back as a `400`).
const MAX_GEOTILE_LEVEL = 29;

// Geotile `z/x/y` → WGS84 `[west, south, east, north]`. The same arithmetic
// as `backend/earthx/catalog/coverage.py::cell_bbox` (kept in lockstep by the
// worked examples in coverage.test.ts) — columns run east from the
// antimeridian, rows run south from the top of the Web-Mercator square.
export function cellBbox(key: string): Bbox {
  const parts = key.split('/');
  if (parts.length !== 3) throw new InvalidCellKey(`grid key "${key}" is not z/x/y`);
  const [z, x, y] = parts.map(Number);
  if (![z, x, y].every(Number.isInteger)) {
    throw new InvalidCellKey(`grid key "${key}" has a part that is not a number`);
  }
  if (z < 0 || z > MAX_GEOTILE_LEVEL) {
    throw new InvalidCellKey(`grid key "${key}" names a level outside 0..${MAX_GEOTILE_LEVEL}`);
  }
  const side = 2 ** z;
  if (x < 0 || x >= side || y < 0 || y >= side) {
    throw new InvalidCellKey(`grid key "${key}" lies outside the grid of its own level`);
  }
  const west = (x / side) * 360 - 180;
  const east = ((x + 1) / side) * 360 - 180;
  const north = mercatorLatitude(y / side);
  const south = mercatorLatitude((y + 1) / side);
  return [west, south, east, north];
}

function mercatorLatitude(fraction: number): number {
  return (Math.atan(Math.sinh(Math.PI * (1 - 2 * fraction))) * 180) / Math.PI;
}

function bboxPolygon([west, south, east, north]: Bbox): GeoJSON.Polygon {
  return {
    type: 'Polygon',
    coordinates: [
      [
        [west, south],
        [east, south],
        [east, north],
        [west, north],
        [west, south],
      ],
    ],
  };
}

// Cells as a GeoJSON `FeatureCollection` for the `fill` layer, `n` carried as
// the property the paint expression reads. A key the arithmetic above cannot
// parse is skipped rather than thrown — one bad cell must not blank the whole
// map (the source is our own backend, but a response is still outside input).
export function cellsToFeatureCollection(cells: CoverageCell[]): GeoJSON.FeatureCollection {
  const features: GeoJSON.Feature[] = [];
  for (const cell of cells) {
    let bbox: Bbox;
    try {
      bbox = cellBbox(cell.k);
    } catch {
      continue;
    }
    features.push({ type: 'Feature', properties: { n: cell.n }, geometry: bboxPolygon(bbox) });
  }
  return { type: 'FeatureCollection', features };
}

// A MapLibre `fill-color` expression, log-scaled, its stops derived from this
// response's own maximum rather than a fixed density (adr/0004 §5: "die
// Stützstellen werden aus dem Maximum der Antwort abgeleitet", replacing the
// prototype's scale that was only ever calibrated to BIOMASS, F12). `n >= 1`
// always holds — a cell only exists because it counted something — so log10
// is always defined and the bottom stop is fixed at 0.
const PALETTE = ['#2c7bb6', '#00a6ca', '#a6d96a', '#fdae61', '#d7191c'];

export function coverageFillColorExpression(maxCount: number): unknown[] {
  const logTop = Math.log10(Math.max(1, maxCount));
  const stops: unknown[] = [];
  // MapLibre refuses an `interpolate` expression whose stops are not
  // *strictly* ascending (found by loading the layer in a real browser, not
  // by a type check: `addLayer` throws at runtime, silently, since its
  // `paint` type is untyped `any`). `maxCount <= 1` collapses every step to
  // the same `logTop * t == 0`, so each stop is nudged past the previous one
  // rather than left equal to it — the color at 0 is what every cell shows
  // either way, since `n === 1` for all of them in that case.
  let previous = -Infinity;
  PALETTE.forEach((color, i) => {
    const t = PALETTE.length === 1 ? 0 : i / (PALETTE.length - 1);
    const value = Math.max(logTop * t, previous + Number.EPSILON);
    stops.push(value, color);
    previous = value;
  });
  return ['interpolate', ['linear'], ['log10', ['get', 'n']], ...stops];
}

// A drawn/uploaded AOI can carry longitudes outside ±180: MapLibre pans
// across repeated world copies at low zoom, so a rectangle dragged over a
// wrapped copy ends up with corners like -200 or 210. The coverage route
// checks ±180 and refuses anything past it (`InvalidCoverageQuery` → `400`),
// which made the heatmap vanish rather than just clamp to the same area a
// non-wrapped view would show. A hard clamp, not an antimeridian split —
// the smallest fix for the reported bug, not a general AOI-longitude fix.
export function clampBboxLongitude([west, south, east, north]: Bbox): Bbox {
  return [Math.max(-180, Math.min(180, west)), south, Math.max(-180, Math.min(180, east)), north];
}

// The zoom brake of `plans/m2-05-coverage.md` §6.8: `footprints_advised`
// alone answers "is the checked total small enough", but a strict filter
// (e.g. a narrow date range) can keep a *world* view under 500 hits, and
// drawing five hundred tiny footprints across the whole globe reads as noise,
// not as scenes. This constant is chosen, not measured — "roughly a country
// is in view" — and kept in one place so a later measurement can move it.
export const FOOTPRINT_MIN_ZOOM = 4;

// The cap on how many footprints M2-07c will actually fetch and draw. Equal
// to the backend's own `FOOTPRINT_THRESHOLD` (`catalog/coverage.py`): once
// `footprints_advised` is true, `total_count` is already known to be below
// it, so this is a ceiling that is never expected to bite, only a guard
// against fetching an unbounded number of pages if it ever is.
export const FOOTPRINT_FETCH_LIMIT = 500;

export function showFootprints(result: CoverageResponse | null, zoom: number): boolean {
  return !!result && result.footprints_advised && zoom >= FOOTPRINT_MIN_ZOOM;
}

// M3-19 (Otto, 23.09.2026, replacing adr/0010 answer 6a): without an AOI the
// heatmap asks only for the visible map extent, and the cell level follows
// the map's zoom so that roughly the same number of cells
// (`TARGET_CELL_COUNT`) covers the screen at any zoom, instead of a level
// derived from zoom alone asking for far fewer cells than the screen has
// room for (zoom 1.6 asked for level 1 — four cells for the whole globe —
// before a fixed z6 cap ever applied). See
// `plans/m3-19-weltueberblick-ausschnitt.md` §1 for the derivation and §2 for
// the measured cell counts/payload sizes it is chosen from. With an AOI nothing
// changes: the request is a real spatial filter, and the map's zoom alone
// still decides (`Math.floor`, in `store.ts`).
//
// A geotile cell of level `L` is `512 * 2^(mapZoom - L)` CSS pixels wide at
// map zoom `mapZoom` (MapLibre draws the whole world at `512 * 2^mapZoom`
// px), so a `width * height` px viewport holds
// `(width * height) / (512 * 2^(mapZoom - L))²` cells — solved for `L` below.
export const TARGET_CELL_COUNT = 2000;
export const MIN_VIEWPORT_LEVEL = 4;

export function levelForViewport(mapZoom: number, width: number, height: number): number {
  if (!Number.isFinite(mapZoom) || !(width > 0) || !(height > 0)) return MIN_VIEWPORT_LEVEL;
  const raw = mapZoom + Math.log2(512 * Math.sqrt(TARGET_CELL_COUNT / (width * height)));
  return Math.max(MIN_VIEWPORT_LEVEL, Math.min(MAX_GEOTILE_LEVEL, Math.round(raw)));
}

// The grid the request bbox is rounded to before it goes out — blocks of
// `2^VIEWPORT_ROUND_LEVELS` × `2^VIEWPORT_ROUND_LEVELS` cells (8×8) of the
// requested level, so a small pan or zoom-and-back keeps asking the very same
// question instead of a new one on every `moveend` (plan §3).
export const VIEWPORT_ROUND_LEVELS = 3;

// Inverse of `mercatorLatitude`: how far down the Web-Mercator square (0 at
// the north edge the grid can draw, 1 at the south edge) a WGS84 latitude
// falls. Clamped to the same ±85.0511° the grid itself is built on — a
// latitude past it would not round-trip through `mercatorLatitude` either.
const MAX_MERCATOR_LATITUDE = 85.0511287798066;

function mercatorFraction(latitude: number): number {
  const clamped = Math.max(-MAX_MERCATOR_LATITUDE, Math.min(MAX_MERCATOR_LATITUDE, latitude));
  const radians = (clamped * Math.PI) / 180;
  return (1 - Math.asinh(Math.tan(radians)) / Math.PI) / 2;
}

// Rounds a viewport extent outward to the geotile block grid of `level`
// (plan §3): a small pan or zoom that stays inside the same block asks for
// exactly the same `bbox` as before, which is what lets `refreshCoverage`
// recognise it as already answered instead of firing a new request.
export function roundBboxToGrid([west, south, east, north]: Bbox, level: number): Bbox {
  const side = 2 ** Math.max(0, Math.floor(level));
  const clampedWest = Math.max(-180, Math.min(180, west));
  const clampedEast = Math.max(-180, Math.min(180, east));
  const colWest = Math.floor(((clampedWest + 180) / 360) * side);
  const colEast = Math.min(side, Math.max(colWest + 1, Math.ceil(((clampedEast + 180) / 360) * side)));
  const rowNorth = Math.floor(mercatorFraction(north) * side);
  const rowSouth = Math.min(side, Math.max(rowNorth + 1, Math.ceil(mercatorFraction(south) * side)));
  return [
    (colWest / side) * 360 - 180,
    mercatorLatitude(rowSouth / side),
    (colEast / side) * 360 - 180,
    mercatorLatitude(rowNorth / side),
  ];
}

// A viewport that crosses ±180° — the antimeridian itself, or a low-zoom view
// repeating world copies (MapLibre pans across those rather than wrapping) —
// is asked for as one band across the full -180..180 width, not a clamped
// slice or two separate requests either of which would misdraw a footprint
// whose centroid sits right at the wrap (plan §3).
export function bandViewportBbox([west, south, east, north]: Bbox): Bbox {
  if (west < -180 || east > 180) return [-180, south, 180, north];
  return [west, south, east, north];
}

// Whether `outer` (a previously requested, rounded bbox) still covers
// `inner` (the current raw viewport) — the reuse check of plan §3: a
// pan/zoom that stays inside what was already asked for needs no new
// request.
export function bboxContains(outer: Bbox, inner: Bbox): boolean {
  return outer[0] <= inner[0] && outer[2] >= inner[2] && outer[1] <= inner[1] && outer[3] >= inner[3];
}

// The legend's completeness line (`plans/m2-05-coverage.md` §3.5 table).
// `null` means the legend needs no addition. English here, like the rest of
// the interface (D25; `adr/0004` §5.3 nachtrag 22.09.2026 lifts the earlier
// German-terms default).
export function completenessNote(result: CoverageResponse): string | null {
  const { counted, total_count: total } = result;
  const table: Record<CoverageCompleteness, () => string | null> = {
    complete: () => null,
    truncated: () => (total === null ? 'truncated' : `showing ${counted} of ${total} scenes`),
    sample: () => (total === null ? 'sample' : `sample: ${counted} of ${total} scenes`),
  };
  return table[result.completeness]();
}

File: backend/earthx/catalog/coverage.py
"""The coverage seam: one question, one answer, whoever answers it (adr/0004 §5).

The question is "how dense is dataset X under filter F on grid level z", the answer
is cells with a count, a time histogram and — the part that makes it trustworthy — a
**checked** completeness. Which of the three ways of adr/0004 §5 answers it is decided
by the registry entry of the dataset, not by the caller: upstream aggregation, SQL in
our own PostGIS, or a declared sample.

Why the seam is here and not in ``adapters``: the grid, the counting rule, the level
caps and rule V are the platform's decisions and hold for every source
(``architekturplan.md`` 3.1, "STAC-Modell, pgstac, Suche"). What a *particular* source
speaks — that Earth Search answers ``/aggregate`` only on GET, what its parameters are
called, where it truncates — stays in ``adapters``. That is also the direction the
import contracts allow: ``adapters`` may import ``catalog``, never the other way
round.

Two properties are deliberate and both are about not claiming more than was measured:

* **``overflow`` is not believed.** Earth Search reports ``overflow: 0`` even where a
  third of the scenes are missing (adr/0004 §3.3 and its note of 20.09.2026). Only
  ``sum(cells) == total_count`` decides, and :func:`check_completeness` is where that
  happens — a function, so it can be tested rather than trusted.
* **A cap clamps, it does not refuse.** Asking for a finer level than a dataset or the
  source can carry returns the coarser map with the level it actually used. The map
  gets coarser, never empty — the same posture as E5 on the answering side.
"""

from __future__ import annotations

import math
from collections.abc import Mapping
from dataclasses import dataclass
from datetime import datetime
from enum import Enum
from typing import Any, Protocol, runtime_checkable

from earthx.catalog.registry import DatasetConfig

# adr/0004 §5, Otto's answer to question 3 of 19.09.2026: below this many hits the
# viewer draws the real footprints instead of a density. It lives here, once, so that
# the rule is the same wherever it is asked and can be checked in pytest; the zoom
# stays the additional brake and sits in the frontend (M2-07c).
FOOTPRINT_THRESHOLD = 500

# adr/0004 §3.3, note of 20.09.2026: without a spatial filter the source truncates
# from z7 on, and the answer passes the ~500 kB mark at which §5 wants to reconsider
# vector tiles. A world view asks z0 to z3 in practice — this only stops someone
# asking for z8 with no bounding box.
WORLD_LEVEL_CAP = 6

# Geotile is the XYZ scheme of the map itself (adr/0004 §3.7), so a level is a zoom.
# Geotile itself runs out here — below a metre a cell is past any imagery — and Earth
# Search happens to draw the same line: a precision outside 0..29 comes back as a 400
# (measured 20.09.2026, plans/m2-05-coverage.md §3.4). The bound is the grid's, the
# measurement only confirms it, so it stays in `catalog` with the rest of the grid.
MAX_GEOTILE_LEVEL = 29

# Measured on 20.09.2026 (plans/m2-05-coverage.md §3.1): Earth Search ignores
# ``datetime_frequency_interval`` — day, month, year and even a nonsense value all
# return the same monthly buckets. So the histogram is monthly, and no caller is
# offered a choice that would not be kept. A finer histogram is a new measurement and
# a decision of its own, not a parameter.
HISTOGRAM_INTERVAL = "month"


class CoverageError(Exception):
    """Base of the errors this seam raises, so a caller can catch the family."""


class InvalidCoverageQuery(CoverageError, ValueError):
    """The request breaks one of our own rules, before anything is sent upstream."""


class CoverageProviderMismatch(CoverageError, LookupError):
    """This dataset is answered by another way than the one that was asked.

    A dispatch mistake, not a user error — the counterpart of ``UnsupportedSource``
    for the search path.
    """


class UpstreamCoverageShapeError(CoverageError, RuntimeError):
    """The source answered something that is not an aggregation we can read."""


class Completeness(Enum):
    """The mandatory field of rule V (adr/0004 §5), in the wording of its note of
    20.09.2026: English on the wire, German in the UI and in the ADR.

    ``complete`` = vollständig, ``truncated`` = gekappt, ``sample`` = Stichprobe.
    """

    COMPLETE = "complete"
    TRUNCATED = "truncated"
    SAMPLE = "sample"


@dataclass(frozen=True, slots=True)
class CoverageCell:
    """One grid cell and how many scenes have their centroid in it.

    The key is the geotile key ``z/x/y``, exactly as the source writes it — there is
    nothing to translate. :func:`cell_bbox` turns it into a box where one is needed;
    on the wire it stays the key, which keeps the world overview at z6 near 160 kB
    instead of the ~570 kB a ready-made polygon would cost (Otto, F4 of 20.09.2026).
    """

    key: str
    count: int


@dataclass(frozen=True, slots=True)
class HistogramBucket:
    """One step of the time histogram, from the same request as the cells."""

    start: datetime
    count: int


@dataclass(frozen=True, slots=True)
class CoverageQuery:
    """What is being asked. Checked here, so an unchecked one cannot exist.

    ``level`` is the geotile level the answer should use. It is *not* clamped in here:
    clamping needs the registry entry, so :func:`level_for_viewport` does it and the
    caller passes the result. What this checks is only what can be wrong on its own.
    """

    dataset_id: str
    level: int
    bbox: tuple[float, float, float, float] | None = None
    intersects: Mapping[str, Any] | None = None
    start: datetime | None = None
    end: datetime | None = None
    max_cloud_cover: float | None = None

    def __post_init__(self) -> None:
        self._check_level()
        self._check_area()
        self._check_time()
        self._check_cloud_cover()

    @property
    def has_spatial_filter(self) -> bool:
        return self.bbox is not None or self.intersects is not None

    @property
    def is_unfiltered(self) -> bool:
        """True for the world overview: no area, no time window, no cloud filter.

        This is the one answer Otto gave the long lifetime on 20.09.2026 (F3): at
        thirty million items a day's new scenes change nothing visible, and it is the
        view every user sees first.
        """
        return not self.has_spatial_filter and self.start is None and self.end is None and self.max_cloud_cover is None

    def _check_level(self) -> None:
        if not 0 <= self.level <= MAX_GEOTILE_LEVEL:
            raise InvalidCoverageQuery(f"grid level {self.level} is outside 0..{MAX_GEOTILE_LEVEL}")

    def _check_area(self) -> None:
        # No message below names a coordinate: an exception text becomes a log line and
        # an error body, and the AOI belongs in neither (projektplan.md 7, point 6).
        if self.bbox is not None and self.intersects is not None:
            raise InvalidCoverageQuery("bbox and intersects ask two different questions; send one")
        if self.bbox is not None:
            if len(self.bbox) != 4:
                raise InvalidCoverageQuery("bbox needs four values: west, south, east, north")
            west, south, east, north = self.bbox
            if not all(-90.0 <= value <= 90.0 for value in (south, north)):
                raise InvalidCoverageQuery("bbox latitudes are outside ±90")
            if not all(-180.0 <= value <= 180.0 for value in (west, east)):
                raise InvalidCoverageQuery("bbox longitudes are outside ±180")
            if south >= north:
                raise InvalidCoverageQuery("bbox is upside down: south is not below north")
        if self.intersects is not None:
            self._check_geometry(self.intersects)

    @staticmethod
    def _check_geometry(geometry: Mapping[str, Any]) -> None:
        """Enough of a check that nothing shapeless reaches a query string.

        Not a full GeoJSON validator — the source is the authority on its own input.
        This refuses what would otherwise be serialised into a URL without anyone ever
        having looked at it.
        """
        if not isinstance(geometry, Mapping):
            raise InvalidCoverageQuery("intersects is not a GeoJSON object")
        kind = geometry.get("type")
        if kind not in {"Polygon", "MultiPolygon"}:
            raise InvalidCoverageQuery("intersects needs a Polygon or MultiPolygon")
        coordinates = geometry.get("coordinates")
        if not isinstance(coordinates, list) or not coordinates:
            raise InvalidCoverageQuery("intersects carries no coordinates")
        if kind == "Polygon":
            rings = coordinates
        else:
            rings = [ring for shape in coordinates if isinstance(shape, list) for ring in shape]
        if not any(_is_ring(ring) for ring in rings):
            # Without this, ``{"coordinates": [[]]}`` passes and is serialised into a
            # query string, and the refusal arrives from the source as a 400 instead of
            # from us before anything is sent.
            raise InvalidCoverageQuery("intersects has no ring with at least four positions")
        for ring in rings:
            if _is_ring(ring):
                _check_ring_bounds(ring)

    def _check_time(self) -> None:
        for name, value in (("start", self.start), ("end", self.end)):
            if value is not None and value.tzinfo is None:
                raise InvalidCoverageQuery(f"{name} has no timezone; STAC instants carry one")
        if self.start is not None and self.end is not None and self.start > self.end:
            raise InvalidCoverageQuery("time window ends before it starts")

    def _check_cloud_cover(self) -> None:
        # The one content filter of M2, listed rather than passed through: a free query
        # parameter would be an open door outwards again, and CQL2 stays off anyway
        # (adr/0005 rule VI).
        if self.max_cloud_cover is None:
            return
        if not 0.0 <= self.max_cloud_cover <= 100.0:
            raise InvalidCoverageQuery("max_cloud_cover is a percentage, so it lies between 0 and 100")


@dataclass(frozen=True, slots=True)
class CoverageResult:
    """The answer, the same shape whichever way produced it (adr/0004 K3)."""

    dataset_id: str
    level: int
    cells: tuple[CoverageCell, ...]
    total_count: int | None
    completeness: Completeness
    histogram: tuple[HistogramBucket, ...]
    from_cache: bool
    # Constant for now, and named rather than assumed: every way of adr/0004 §5 counts
    # a scene into exactly one cell, and the legend says so ("Aufnahmen mit Mittelpunkt
    # in der Zelle"). A second grid or a second counting rule would have to say so here.
    grid: str = "geotile"
    counting: str = "centroid"
    # Set only by :func:`extent_result` and the ``local-sql`` area way
    # (`catalog.local_coverage.area_coverage`, M3-11c): a one-off product has
    # nothing to count, so it answers with the ground it covers instead of a
    # density (adr/0004 §5, "Wo welcher Teil liegt" / "Einmal-Produkte"). For the
    # area way this is the bbox of ``area`` below, not the collection's own extent.
    extent: tuple[float, float, float, float] | None = None
    # Set only by the ``local-sql`` area way (M3-11c, adr/0009 §6 F5): the union of
    # a materialized one-off product's own item footprints, as a GeoJSON
    # ``MultiPolygon`` — the honest coverage of a dataset with nothing to count,
    # cropped to whatever spatial filter the query carried. ``None`` everywhere
    # else, including the ``extent_result`` way, which answers a different question
    # (the collection's declared extent, not a union of items).
    area: Mapping[str, Any] | None = None
    # Filters the query carried that this answer could not honour and silently
    # dropped rather than reject (Otto, 26.09.2026, M3-11c): ``"datetime"`` for a
    # dataset with ``capabilities.time_range=False``, ``"max_cloud_cover"`` on the
    # area way, which has no cloud cover to filter on. Named, not just dropped, so
    # the frontend can say so instead of the map quietly answering a narrower
    # question than the one asked (the same posture as `completeness=truncated`).
    ignored_filters: tuple[str, ...] = ()

    @property
    def counted(self) -> int:
        """What the cells add up to — derived, never passed in.

        Rule V lives on this number being the cells' own sum. As a field it could be
        set to something the cells do not say, and the check would then compare the
        total against a number with no relation to the map.
        """
        return sum(cell.count for cell in self.cells)

    @property
    def max_count(self) -> int:
        """The largest cell, which is where the frontend anchors its log scale."""
        return max((cell.count for cell in self.cells), default=0)

    @property
    def footprints_advised(self) -> bool:
        """Whether to draw real footprints instead of the density (adr/0004 §5).

        ``total_count is None`` — a declared sample — deliberately keeps the density:
        the replacement rule for M2-07c says footprints must not appear automatically
        where there is no checked total to compare against.
        """
        return self.total_count is not None and self.total_count < FOOTPRINT_THRESHOLD


class CoverageSource(Protocol):
    """Whoever answers the question, answers exactly this.

    Only the query and the dataset entry, nothing else: the three ways of adr/0004 §5
    need different things to do their work — the upstream way a gateway and a cache,
    the local one a database connection — and a seam that named any of them would be
    the shape of one way rather than of the question. Whoever assembles the parts binds
    them before handing the source over, which is what ``api`` does for the federated
    search already.
    """

    async def __call__(self, query: CoverageQuery, config: DatasetConfig) -> CoverageResult:
        """Density, histogram and a checked completeness for this query."""
        ...


@runtime_checkable
class CoverageCache(Protocol):
    """A key-value store with an expiry. Missing is normal; failing is survivable.

    The same shape as ``earthx.adapters.cache.SearchCache`` and satisfied by the same
    ``PostgresSearchCache``. It is declared a second time rather than imported because
    the two point in opposite directions: ``SearchCache`` is what ``adapters`` asks of
    ``catalog``, this is what the seam in ``catalog`` asks of whoever answers it, and
    ``catalog`` must not import ``adapters`` (architekturplan.md 3.1).
    """

    async def get(self, key: str) -> dict[str, Any] | None:
        """The stored value, or None if it is absent or expired."""
        ...

    async def set(self, key: str, value: dict[str, Any], *, ttl_s: float, dataset_id: str) -> None:
        """Store a value for ``ttl_s`` seconds."""
        ...


def check_completeness(
    counted: int,
    total_count: int | None,
    *,
    simplified_aoi: bool = False,
    sampled: bool = False,
) -> Completeness:
    """Rule V of adr/0004 §5 as a function, so it is checked and not intended.

    Four cases, and three of them are the ones that cost a map its honesty:

    * a declared sample is a sample, whatever the numbers say;
    * a simplified or boxed AOI answers a different question than the one asked
      (adr/0004 §3.4), so the answer is truncated even when the sums agree;
    * **without a ``total_count`` there is nothing to compare against**, and a source
      that cannot state a total cannot prove completeness (note of 20.09.2026);
    * otherwise it is complete exactly when the cells add up to the total. The source's
      own ``overflow`` never enters into it — it reports 0 while a third of the scenes
      are missing (adr/0004 §3.3).
    """
    if sampled:
        return Completeness.SAMPLE
    if simplified_aoi or total_count is None or counted != total_count:
        return Completeness.TRUNCATED
    return Completeness.COMPLETE


def extent_result(dataset_id: str, bbox: tuple[float, float, float, float]) -> CoverageResult:
    """The answer for a one-off product: its extent, not a density (adr/0004 §5).

    Checked *before* a provider is even asked (``registry.py`` already forbids the
    combination of ``single_coverage_product`` with upstream aggregation) — a single
    coverage has nothing to count, so there is no cell, no histogram, and its
    completeness is trivially whole: the extent is what the collection states,
    nothing is being compared against a total that does not apply here.
    """
    return CoverageResult(
        dataset_id=dataset_id,
        level=0,
        cells=(),
        total_count=None,
        completeness=Completeness.COMPLETE,
        histogram=(),
        from_cache=False,
        extent=bbox,
    )


def level_for_viewport(zoom: int, config: DatasetConfig, *, has_spatial_filter: bool) -> int:
    """The grid level to ask for: the map's zoom under both caps of adr/0004 §5.

    Clamped, never refused — a level finer than the caps allow returns the coarser map
    with the level it used, and the caller reports that level in the answer.

    Cap one is the dataset's own (``max_geotile_level``): once a cell is smaller than a
    footprint, the centroid rule draws a dot pattern instead of a coverage. Cap two is
    the world cap of the note of 20.09.2026: without a spatial filter the source
    truncates from z7 on anyway.
    """
    level = max(0, min(int(zoom), MAX_GEOTILE_LEVEL))
    level = min(level, config.coverage.max_geotile_level)
    if not has_spatial_filter:
        level = min(level, WORLD_LEVEL_CAP)
    return level


def parse_cell_key(key: str) -> tuple[int, int, int]:
    """Read a geotile key ``z/x/y``, or say why it is not one.

    Strict on purpose: the key comes from outside and ends up in a map layer. A key
    whose column or row lies outside ``2**z`` is not a cell that exists.
    """
    parts = key.split("/")
    if len(parts) != 3:
        raise UpstreamCoverageShapeError(f"grid key {key!r} is not z/x/y")
    try:
        level, column, row = (int(part) for part in parts)
    except ValueError:
        raise UpstreamCoverageShapeError(f"grid key {key!r} has a part that is not a number") from None
    if not 0 <= level <= MAX_GEOTILE_LEVEL:
        raise UpstreamCoverageShapeError(f"grid key {key!r} names a level outside 0..{MAX_GEOTILE_LEVEL}")
    side = 1 << level
    if not (0 <= column < side and 0 <= row < side):
        raise UpstreamCoverageShapeError(f"grid key {key!r} lies outside the grid of its own level")
    return level, column, row


def cell_bbox(key: str) -> tuple[float, float, float, float]:
    """The cell as west, south, east, north in WGS84.

    Geotile is the ordinary XYZ scheme: columns run east from the antimeridian, rows
    run south from the top of the Web-Mercator square. Kept here rather than in the
    frontend or the adapter because both the local SQL way and the tests need the very
    same arithmetic (adr/0004 §3.7).
    """
    level, column, row = parse_cell_key(key)
    side = 1 << level
    west = column / side * 360.0 - 180.0
    east = (column + 1) / side * 360.0 - 180.0
    north = _mercator_latitude(row / side)
    south = _mercator_latitude((row + 1) / side)
    return (west, south, east, north)


def geotile_key(longitude: float, latitude: float, level: int) -> str:
    """The geotile cell whose area contains a WGS84 point — the inverse of
    :func:`cell_bbox`.

    For a coverage way that rasterizes itself instead of asking a source to
    (adr/0004 §5, Option 6, the declared-sample way of M2-09b-3): a footprint is
    counted into the one cell its centroid falls in, by the same grid and the
    same centroid rule every other way uses — this is that arithmetic, kept here
    rather than in `adapters` because it is the grid's own definition, not
    anything about a particular source (module docstring above).

    A point at the edge of the grid — past ±180° longitude, or past the Mercator
    limit near a pole — is clamped into the nearest cell rather than refused: a
    measured centroid is never truly outside the world it came from, only
    outside the numeric range this projection can draw, and refusing it would
    turn one odd footprint into a failed request instead of one cell that is a
    little off.
    """
    side = 1 << level
    column = min(side - 1, max(0, int((_clamped_longitude(longitude) + 180.0) / 360.0 * side)))
    row = min(side - 1, max(0, int(_mercator_row(latitude) * side)))
    return f"{level}/{column}/{row}"


# atan(sinh(pi)) in degrees — where `_mercator_latitude`'s row 0 already lands, so the
# forward mapping below clamps to exactly the range the grid can already draw.
_MAX_MERCATOR_LATITUDE = 85.05112877980659


def _clamped_longitude(longitude: float) -> float:
    return max(-180.0, min(180.0, longitude))


def _clamped_latitude(latitude: float) -> float:
    return max(-_MAX_MERCATOR_LATITUDE, min(_MAX_MERCATOR_LATITUDE, latitude))


def _mercator_row(latitude: float) -> float:
    """Fraction of the way down the Web-Mercator square — the inverse of
    `_mercator_latitude`, by solving its formula for `fraction`."""
    radians = math.radians(_clamped_latitude(latitude))
    return (1.0 - math.asinh(math.tan(radians)) / math.pi) / 2.0


def _check_ring_bounds(ring: list[Any]) -> None:
    """Every corner of a ring lies on the globe.

    Otherwise a polygon with a longitude of 999 and a latitude of 888 is accepted
    (the source answers ``200`` and a plausible-looking number instead of refusing
    it — adr/0005 §3.5 describes the same silent acceptance for a bbox). None of the
    values enters the message (projektplan.md 7, point 6).
    """
    for longitude, latitude, *_rest in ring:
        if not -180.0 <= float(longitude) <= 180.0:
            raise InvalidCoverageQuery("intersects longitude is outside ±180")
        if not -90.0 <= float(latitude) <= 90.0:
            raise InvalidCoverageQuery("intersects latitude is outside ±90")


def _is_ring(ring: Any) -> bool:
    """A closed outline needs four positions of two numbers, three of them distinct."""
    return (
        isinstance(ring, list)
        and len(ring) >= 4
        and all(isinstance(point, (list, tuple)) and len(point) >= 2 for point in ring)
        and all(_is_number(value) for point in ring for value in point[:2])
    )


def _is_number(value: Any) -> bool:
    """A coordinate, and not a bool — ``True`` is an ``int`` and would pass otherwise."""
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def _mercator_latitude(fraction: float) -> float:
    """Latitude at ``fraction`` of the way down the Web-Mercator square."""
    return math.degrees(math.atan(math.sinh(math.pi * (1.0 - 2.0 * fraction))))
